_posts/2020-12-20-rocksdb.md

```diff
@@ -0,0 +1,121 @@
+
+# How to use RocksDBStateBackend
+
+RocksDB is fully embedded within and fully managed by the TaskManager process. RocksDBStateBackend can be configured at the cluster level as the default for the entire cluster, or at the job level for individual jobs. The job level configuration takes precedence over the cluster level configuration.
+
+## Cluster Level
+
+Add the following configuration in [`conf/flink-conf.yaml`](https://ci.apache.org/projects/flink/flink-docs-stable/deployment/config.html):
+
+    state.backend: rocksdb
+    state.backend.incremental: true
+    state.checkpoints.dir: hdfs:///flink-checkpoints # location to store checkpoints
+
+# Job Level
+
+Add the following into your job’s code after StreamExecutionEnvironment is created:
+
+    # 'env' is the created StreamExecutionEnvironment
+    # 'true' is to enable incremental checkpointing
+    env.setStateBackend(new RocksDBStateBackend("hdfs:///fink-checkpoints", true));   
+
+<div class="alert alert-info" markdown="1">
+<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Note</span>
+In addition to HDFS, you can also use other on-premises or cloud based object stores if the corresponding dependencies are added under [FLINK_HOME/plugins](https://ci.apache.org/projects/flink/flink-docs-stable/deployment/filesystems/plugins.html).
+</div>     
+
+# Best Practices and Advanced Configuration
+
+We hope this overview helped you gain a better understanding of the role of RocksDB in Flink and how to successfully run a job with RocksDBStateBackend. To round it off, we’ll explore some best practices and a few reference points for further troubleshooting and performance tuning.
+
+## State Location in RocksDB
+
+As mentioned earlier, in-flight state in RocksDBStateBackend is spilled to files on disks. These files are located under the directory specified by the Flink configuration [`state.backend.rocksdb.localdir`](https://ci.apache.org/projects/flink/flink-docs-stable/deployment/config.html#state-backend-rocksdb-localdir). Because disk performance has a direct impact on RocksDB’s performance, it’s recommended that this directory is located on a **local** disk.  It’s discouraged to configure it to a remote network-based location like NFS or HDFS, as writing to remote disks is usually slower. Also high availability is not a requirement for in-flight state. Local SSD disks are preferred if high disk throughput is required.
```

Comment:
```suggestion
As mentioned earlier, in-flight state in RocksDBStateBackend is spilled to files on disk. These files are located under the directory specified by the Flink configuration [`state.backend.rocksdb.localdir`](https://ci.apache.org/projects/flink/flink-docs-stable/deployment/config.html#state-backend-rocksdb-localdir). Because disk performance has a direct impact on RocksDB’s performance, it’s recommended that this directory is located on a **local** disk.  It’s discouraged to configure it to a remote network-based location like NFS or HDFS, as writing to remote disks is usually slower. Also high availability is not a requirement for in-flight state. Local SSD disks are preferred if high disk throughput is required.
```